compose: Update the preview after restoring a draft.

skunkmb opened the pull request:
This PR attempts to fix #5951 by re-opening the preview view after a draft was opened, if it was open already.

I also thought that re-opening the editing view after opening a draft could make sense, [as suggested by @timabbott](https://github.com/zulip/zulip/issues/5951#issuecomment-318728809). Please let me know if instead of conditionally opening the preview, I should change it so that the editor is always opened.

(Done for the *"Issue #5951 Restoring message draft doesn't update rendered preview"* Google Code-in task.)

skunkmb commented:
It seems like the error that's causing the tests to fail,

```
Exception: Page not found: <404 https://docs.docker.com/engine/understanding-docker/>
```

is also causing other recent commits to fail, like [the test for `c2a41c2`](https://travis-ci.org/zulip/zulip/jobs/319180999). I don't the failing has to do with the content of this pull request, but correct me if I'm wrong.

sampritipanda commented:
I'd say closing the preview and opening the editor would be more sensible. I mean, you restore a draft to edit and send it right, you can see the preview in the drafts window anyway.

skunkmb commented:
@sampritipanda OK, I changed it so that now it just goes straight to the editor. I also kept the `is_preview_shown` method that I originally created and used in the first version. Although it's not being used anymore, I thought it might be useful in the future. I can remove it if you don't think it's helpful, though.

I also added tests to Casper to make sure the returning behavior works as expected.

sampritipanda commented on static/js/drafts.js:
```diff
@@ -148,6 +148,9 @@ exports.restore_draft = function (draft_id) {
     compose_ui.autosize_textarea();
+
+    compose.clear_preview_area();
+
```
I know this sounds really stupid, but what about having this command right below the `clear_compose` command. Like we do all the tear-down first, and then we restore the draft :stuck_out_tongue:

skunkmb replied:
@sampritipanda Sure! That makes sense. I made that changed and pushed it.

sampritipanda commented on frontend_tests/casper_tests/14-drafts.js:
```diff
@@ -43,6 +43,15 @@ casper.then(function () {
+        casper.click('#markdown_preview');
+    });
+});
+
```
So having this action here brings out an un-intended effect which is that pressing c or C to open the create a stream/private message doesn't close the preview which we don't want to depend on for this test to pass I guess? We should probably move this to just above the clicking on the drafts link.

sampritipanda replied:
Oh and, the preview gets closed by the `casper.click("#compose_close");` that is run after creating the drafts, so it doesn't really do it's job :stuck_out_tongue:

sampritipanda commented:
About the `is_preview_shown` thing, I would waver towards not having it because why introduce unnecessary code when it's not used. + it reduces coverage too, so you will be writing a test for the method which isn't yet.

skunkmb commented:
@sampritipanda OK, I removed the `is_preview_shown`. I also moved the test up like you suggested, and made it use the button for a new topic instead of the keyboard shortcut because it is more self-explanatory.

sampritipanda commented:
Yeah great @skunkmb, can you squash the commits into one keeping the commit message of the first commit?

sampritipanda commented:
LGTM otherwise!

skunkmb commented:
@sampritipanda I squashed it. Thanks!

skunkmb commented:
Is there anything else I should fix?

sampritipanda commented:
@skunkmb I don't have merge privileges, but I've approved the task since it looks good enough. Steve will take a look and merge if it looks good, if not he'll add a comment.

skunkmb commented:
@sampritipanda OK, thank you!

showell commented:
Merged!  Thanks @skunkmb @sampritipanda
